### PR TITLE
Preventing port 53 being added as lb rule when dns service is availab…

### PR DESCRIPTION
**Pull request by Spaceman1984**
…le in network offering

## Description

Throwing an error when port 53 is added to a load balancer when DNS is available on the network service offering.

Fixes: #4285 
## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## How Has This Been Tested?

Try to add port 53 as a public port as a load balancer rule, an error will be returned.

### CODE REVIEW

**Comment by Spaceman1984** [comment]
@blueorangutan package

**Comment by Spaceman1984** [comment]
@blueorangutan test

File: server/src/main/java/com/cloud/network/lb/LoadBalancingRulesManagerImpl.java [hunk 1]
```diff
@@ -1598,65 +1602,73 @@ public LoadBalancer createPublicLoadBalancerRule(String xId, String name, String
...
-                if (systemIp != null) {
-                    ipVO = _ipAddressDao.findById(systemIp.getId());
+        IpAddress systemIp = null;
+        NetworkOffering off = _entityMgr.findById(NetworkOffering.class, network.getNetworkOfferingId());
+
+        if (srcPortStart == 53) {
+            List<NetworkOfferingServiceMapVO> offerinServices = _networkOfferingServiceDao.listByNetworkOfferingId(network.getNetworkOfferingId());
+            for (NetworkOfferingServiceMapVO serviceMapVo: offerinServices){
+                if (serviceMapVo.getService() == Service.Dns.getName()) {
+                    throw new InvalidParameterValueException("Error adding load balancer rule, cannot add port 53 with network service offering having DNS service.");
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/network/lb/LoadBalancingRulesManagerImpl.java:1612]: Loads of false positives on the diff. This is actually the only change @Spaceman1984 , or am I missing something else?

Comment (Spaceman1984) [reply server/src/main/java/com/cloud/network/lb/LoadBalancingRulesManagerImpl.java:1612]: Yes @DaanHoogland , mostly reformatting, there was an unnecessary check which I removed:
```
 LoadBalancer result = null;
 if (result == null) {
```

**Comment by Spaceman1984** [comment]
@blueorangutan test

File: server/src/main/java/com/cloud/network/lb/LoadBalancingRulesManagerImpl.java [hunk 2]
```diff
@@ -1598,65 +1602,73 @@ public LoadBalancer createPublicLoadBalancerRule(String xId, String name, String
...
-            IpAddress systemIp = null;
-            NetworkOffering off = _entityMgr.findById(NetworkOffering.class, network.getNetworkOfferingId());
-            if (off.isElasticLb() && ipVO == null && network.getVpcId() == null) {
-                systemIp = _ipAddrMgr.assignSystemIp(networkId, lbOwner, true, false);
-                if (systemIp != null) {
-                    ipVO = _ipAddressDao.findById(systemIp.getId());
+        IpAddress systemIp = null;
+        NetworkOffering off = _entityMgr.findById(NetworkOffering.class, network.getNetworkOfferingId());
+
+        if (srcPortStart == 53) {
```
Comment (weizhouapache) [inline server/src/main/java/com/cloud/network/lb/LoadBalancingRulesManagerImpl.java:1608]: @Spaceman1984 
(1) the issue happens only if lb uses source nat IP. so IP needs to be checked
(2) what about port 8081 which is used for haproxy stats ?

Comment (Spaceman1984) [reply server/src/main/java/com/cloud/network/lb/LoadBalancingRulesManagerImpl.java:1608]: I'll look into it @weizhouapache.

Comment (Spaceman1984) [reply server/src/main/java/com/cloud/network/lb/LoadBalancingRulesManagerImpl.java:1608]: I tested port 8081 and it threw an error, seems like there is already some code somewhere that checks for certain ports, I'll see if I should move my code there as well.

Comment (Spaceman1984) [reply server/src/main/java/com/cloud/network/lb/LoadBalancingRulesManagerImpl.java]: I think the code is fine where it is, I've added a check for Source NAT.

**Comment by Spaceman1984** [comment]
@blueorangutan package

**Comment by Spaceman1984** [comment]
@blueorangutan test

**Review by weizhouapache — APPROVED** [review]
lgtm
would it be better to fix it in 4.14 as well ?

**Comment by DaanHoogland** [comment]
@Spaceman1984 can you rebase on 4.14?

**Comment by Spaceman1984** [comment]
Sure @DaanHoogland, I'll try to get to it later today.

**Comment by Spaceman1984** [comment]
Closing in favor of #4442

**Pull request closed without merging**